keps/sig-apps/592-ttl-after-finish/README.md

```diff
@@ -250,17 +257,141 @@ Mitigations:
 
 ## Graduation Criteria
 
-We want to implement this feature for Pods/Jobs first to gather feedback, and
-decide whether to generalize it to custom resources. This feature can be
-promoted to beta after we finalize the decision for whether to generalize it or
-not, and when it satisfies users' need for cleaning up finished resource
-objects, without regressions.
+- The feature implemented for Job, as future work, it can be extended to Pods and perhaps custom resources, but that should happen under separate feature flags.
+- Add necessary tests
+- Graduate to Beta in v1.21
+- Graduate to GA in 1.23
 
-This will be promoted to GA once it's gone a sufficient amount of time as beta
-with no changes. 
 
 [umbrella issues]: https://github.com/kubernetes/kubernetes/issues/42752
 
-## Implementation History
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+* **How can this feature be enabled / disabled in a live cluster?**
+  - [x] Feature gate (also fill in values in `kep.yaml`)
+    - Feature gate name: TTLAfterFinished
+    - Components depending on the feature gate: kube-apiserver, kube-controller-manager
+  - [ ] Other
+    - Describe the mechanism:
+    - Will enabling / disabling the feature require downtime of the control
+      plane?
+    - Will enabling / disabling the feature require downtime or reprovisioning
+      of a node? (Do not assume `Dynamic Kubelet Config` feature is enabled).
+
+* **Does enabling the feature change any default behavior?**
+  No.
+
+* **Can the feature be disabled once it has been enabled (i.e. can we roll back
+  the enablement)?**
+  Yes.
+
+* **What happens if we reenable the feature if it was previously rolled back?**
+  It should work as expected.
+
+* **Are there any tests for feature enablement/disablement?**
+  No.
+
+### Rollout, Upgrade and Rollback Planning
+
+* **How can a rollout fail? Can it impact already running workloads?**
+  It shouldn't impact already running workloads. This is an opt-in feature since users need to 
+  explicitly set the TTLSecondsAfterFinished parameter in the job spec, which is accepted by the 
+  api-server only if the feature is enabled.
+
+* **What specific metrics should inform a rollback?**
+  Unexpected restarts of kube-controller-manager
```

Comment:
How about kube-apiserver?
I suggest also extended 4xx/5xx on Jobs from kube-apiserver (similarly to what I've been pointing out in #2199 that I mentioned above).

Reply:
is that a metric that already exists? if so, what is the name?

Reply:
apiserver_request_total has a "code" (being http code returned to the user):
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go#L76

Reply:
you mean 4xx/5xx replies when reaching to the jobs endpoint on the api-server? why would that be relevant to this feature?

Reply:
They don't have to be relevant. But they may potentially be, e.g. imagine tha tthe conversion function is returning an error if that field is set - that would exactly result in returning 5xx from it.

Reply:
done.